index.js

```diff
@@ -10,125 +10,118 @@ import {
-  switch (selectedValue) {
-    case COMPLETE:
-      renderTasksByStatus(true);
-      break;
-    case INCOMPLETE:
-      renderTasksByStatus(false);
-      break;
-    default:
-      renderTodoList();
-  }
 
-  const tabs = document.querySelectorAll(".filter-tab");
+let searchText = "";
+let auxilaryTodos;
+let filterStatus;
+let searchResult;
+let currentPage = 1;
+const taskPerPage = 3;
+let totalPageCount;
+const startIndex = 0;
+
+export const isFilterTabDisable = () => {
+  todos?.length
+    ? tabs.forEach((tab) => {
+        tab.removeAttribute("disabled");
+      })
+    : tabs.forEach((tab) => {
+        tab.setAttribute("disabled", true);
+        tab.classList.remove("selected");
+      });
+};
 
-  tabs.forEach((tab) => tab.classList.remove("selected"));
+const filteredByPaginate = (todos) => {
+  totalPageCount = Math.ceil(todos.length / taskPerPage);
+  renderLoadMoreButton(todos);
+  return todos.slice(startIndex, taskPerPage * currentPage);
+};
 
-  event.target.classList.add("selected");
+const handleLoadMoreTask = () => {
+  if (todos && currentPage < totalPageCount) currentPage++;
+  renderTodoList(todos);
 };
 
-const renderTasksByStatus = (isDone) => {
-  keepCard1Elements();
-  filteredTodos = filteredTasks.length
-    ? filteredTasks.filter((todo) => todo.isDone === isDone)
-    : todos.filter((todo) => todo.isDone === isDone);
+export const renderLoadMoreButton = (todos) => {
+  if (todos.length >= taskPerPage && currentPage !== totalPageCount) {
+    return (
+      loadMoreButton.classList.remove("hide"),
+      loadMoreButton.classList.add("show")
+    );
+  }
 
-  filteredTodos.map((todo) => {
-    const newTask = createTaskElement(todo);
-    taskList.insertBefore(newTask, taskList.children[1]);
-  });
+  return (
+    loadMoreButton.classList.remove("show"),
+    loadMoreButton.classList.add("hide")
+  );
 };
 
-export const keepCard1Elements = () => {
-  const elementsToKeep = [];
-
-  for (let i = 0; i < taskList.children.length; i++) {
-    if (taskList.children[i].id === "card1") {
-      elementsToKeep.push(taskList.children[i]);
-    }
+const filteredByFilter = (status = "all", todos) => {
+  if (status === "all") {
+    return todos;
+  } else if (status === "complete") {
+    return todos?.filter((todo) => todo.isDone === true);
+  } else if (status === "incomplete") {
+    return todos?.filter((todo) => todo.isDone === false);
   }
+};
 
-  taskList.innerHTML = "";
-
-  elementsToKeep.map((element) => {
-    taskList.appendChild(element);
-  });
+const filteredBySearch = (searchText, todos) => {
+  return todos?.filter((task) => task.value.toLowerCase().includes(searchText));
 };
 
-const debounce = (handleSearch, delay) => {
+const debounce = (filteredBySearch, delay) => {
   let timer;
-
-  return function (...args) {
+  return function () {
     clearTimeout(timer);
-
-    timer = setTimeout(() => {
-      handleSearch(...args);
-    }, delay);
+    return new Promise((resolve) => {
```

Comment:
Why is this returning `Promise`?

Reply:
??

Reply:
To handle the result asynchronously

> ??



> Why is this returning `Promise`?

Reply:
Debouncing has nothing to with asynchronous programming. 
Please read this article to understand better:  https://www.freecodecamp.org/news/javascript-debounce-example/